### PR TITLE
State fixed-size array: `genesis_validators_root`

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Refactoring

**What does this PR do? Why is it needed?**

Replaces `genesis_validators_root` proto definition in the state from `[]byte` to `[32]byte`. 

**Which issues(s) does this PR fix?**

Part of #9820 

**Other notes for review**

`go build ./...` is successful.

### CODE REVIEW

File: deps.bzl [hunk 1]
```diff
@@ -2979,8 +2979,8 @@ def prysm_deps():
     go_repository(
         name = "com_github_prysmaticlabs_protoc_gen_go_cast",
         importpath = "github.com/prysmaticlabs/protoc-gen-go-cast",
-        sum = "h1:k7CCMwN7VooQ7GhfySnaVyI4/9+QbhJTdasoC6VOZOI=",
-        version = "v0.0.0-20210504233148-1e141af6a0a1",
+        sum = "h1:axXDYjfo67qxzP3YldNmDLoGV4dtm2TwfhVgMx6PnTw=",
+        version = "v0.0.0-20211029152443-a17a2c381519",
```
Comment (rkapka) [inline deps.bzl:2983]: I needed to make some changes to our compiler plugin for fixed-size array support.

File: beacon-chain/state/custom-types/BUILD.bazel [hunk 2]
```diff
@@ -4,7 +4,7 @@ go_library(
     name = "go_default_library",
     srcs = ["state_arrays.go"],
     importpath = "github.com/prysmaticlabs/prysm/beacon-chain/state/custom-types",
-    visibility = ["//beacon-chain:__subpackages__"],
+    visibility = ["//visibility:public"],
```
Comment (rkapka) [inline beacon-chain/state/custom-types/BUILD.bazel:7]: Required by SSZ.

File: beacon-chain/state/v1/getters_misc.go [hunk 3]
```diff
@@ -36,12 +36,9 @@ func (b *BeaconState) genesisTime() uint64 {
...
 // GenesisValidatorRoot of the beacon state.
-func (b *BeaconState) GenesisValidatorRoot() []byte {
+func (b *BeaconState) GenesisValidatorRoot() [32]byte {
 	if !b.hasInnerState() {
-		return nil
-	}
-	if b.state.GenesisValidatorsRoot == nil {
-		return params.BeaconConfig().ZeroHash[:]
+		return params.BeaconConfig().ZeroHash
 	}
```
Comment (rkapka) [inline beacon-chain/state/v1/getters_misc.go:42]: Is this change safe to make?

File: network/forks/fork.go [hunk 4]
```diff
@@ -56,14 +53,11 @@ func ForkDigestFromEpoch(currentEpoch types.Epoch, genesisValidatorsRoot []byte)
...
 	genesisTime time.Time,
-	genesisValidatorsRoot []byte,
+	genesisValidatorsRoot [32]byte,
 ) ([4]byte, error) {
 	if genesisTime.IsZero() {
 		return [4]byte{}, errors.New("genesis time is not set")
 	}
-	if len(genesisValidatorsRoot) == 0 {
-		return [4]byte{}, errors.New("genesis validators root is not set")
-	}
```
Comment (rkapka) [inline network/forks/fork.go:66]: This should rather be changed to comparison against zero hash, right?

File: encoding/bytesutil/bytes.go [hunk 5]
```diff
@@ -216,6 +216,14 @@ func SafeCopyBytes(cp []byte) []byte {
 	return nil
 }
 
+func SafeCopy32Bytes(cp [32]byte) [32]byte {
```
Comment (nisdas) [inline encoding/bytesutil/bytes.go:219]: an array is coped by value. so doing
```
newArr := arr
```
works

File: network/forks/fork.go [hunk 6]
```diff
@@ -40,10 +40,7 @@ func IsForkNextEpoch(genesisTime time.Time, genesisValidatorsRoot []byte) (bool,
 
 // ForkDigestFromEpoch retrieves the fork digest from the current schedule determined
 // by the provided epoch.
-func ForkDigestFromEpoch(currentEpoch types.Epoch, genesisValidatorsRoot []byte) ([4]byte, error) {
-	if len(genesisValidatorsRoot) == 0 {
```
Comment (nisdas) [inline network/forks/fork.go:44]: this is probably going to break a ton of tests, you would be better of removing it here

File: network/forks/fork.go (hunk 4 above)
Comment (nisdas) [reply network/forks/fork.go:66]: same applies here

File: network/forks/fork.go (hunk 6 above)
Comment (rkapka) [reply network/forks/fork.go:44]: I did remove the check. Did you mean something else?

**Pull request merged**